File: pronacej-backend-seguridad/src/main/java/net/latinus/sistema/integral/gestion/seguridad/service/general/EncuestaService.java
package net.latinus.sistema.integral.gestion.seguridad.service.general;

import jakarta.servlet.http.HttpServletRequest;
import net.latinus.sistema.integral.gestion.seguridad.model.both.BodyEncriptado;
import net.latinus.sistema.integral.gestion.seguridad.model.both.DocumentoDTO;
import net.latinus.sistema.integral.gestion.seguridad.model.both.encuesta.EncabezadoDTO;
import net.latinus.sistema.integral.gestion.seguridad.model.both.encuesta.EncuestaDTO;
import net.latinus.sistema.integral.gestion.seguridad.model.response.PaginacionResponse;
import net.latinus.sistema.integral.gestion.seguridad.model.response.RespuestaPorDefectoAuditoria;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

public interface EncuestaService {
    // region Encuesta

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data EncuestaDTO
     *
     * @param httpServletRequest request peticion.
     * @param bodyEncriptado request peticion.
     * @return RespuestaPorDefectoAuditoria<EncuestaDTO>
     */
    RespuestaPorDefectoAuditoria<PaginacionResponse<EncuestaDTO>> obtenerListaEncuestas(HttpServletRequest httpServletRequest,
                                                                                        BodyEncriptado bodyEncriptado);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data EncuestaDTO
     *
     * @param httpServletRequest request peticion.
     * @param bodyEncriptado request peticion.
     * @return RespuestaPorDefectoAuditoria<EncuestaDTO>
     */
    RespuestaPorDefectoAuditoria<List<EncuestaDTO>> obtenerEncuestas(HttpServletRequest httpServletRequest,
                                                                     BodyEncriptado bodyEncriptado);

        /**
         * Devuelve un objeto RespuestaPorDefectoAuditoria con data EncabezadoDTO
         *
         * @param httpServletRequest request peticion.
         * @param bodyEncriptado     objeto body encriptado.
         * @return RespuestaPorDefectoAuditoria<PaginacionResponse<EncabezadoDTO>>
         */
    public RespuestaPorDefectoAuditoria<PaginacionResponse<EncabezadoDTO>> obtenerEvaluacionesPorFichaIdentificacion(HttpServletRequest httpServletRequest,
                                                                                                                 BodyEncriptado bodyEncriptado, String nemonicoCategoria);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data EncabezadoDTO
     *
     * @param httpServletRequest request peticion.
     * @param bodyEncriptado     objeto body encriptado.
     * @param nemonicoEncuesta ficha identificacion.
     * @return RespuestaPorDefectoAuditoria<PaginacionResponse<EncabezadoDTO>>
     */
    public RespuestaPorDefectoAuditoria<PaginacionResponse<EncabezadoDTO>> obtenerEvaluacionesPorNemonicoEncuesta(HttpServletRequest httpServletRequest,
                                                                                                                 BodyEncriptado bodyEncriptado, String nemonicoEncuesta);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data EncabezadoDTO
     *
     * @param httpServletRequest request peticion.
     * @param bodyEncriptado     objeto body encriptado.
     * @param nemonicosCategoria lista de nemonicos a ser incluidos en la respuesta.
     * @return RespuestaPorDefectoAuditoria<PaginacionResponse<EncabezadoDTO>>
     */
    public RespuestaPorDefectoAuditoria<PaginacionResponse<EncabezadoDTO>> obtenerEvaluacionesPorNemonicoCategoria(HttpServletRequest httpServletRequest,
                                                                                                                  BodyEncriptado bodyEncriptado, String nemonicoCentro, List<String> nemonicosCategoria );


    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data EncuestaDTO
     *
     * @param httpServletRequest request peticion.
     * @return RespuestaPorDefectoAuditoria<EncuestaDTO>
     */
    RespuestaPorDefectoAuditoria<EncuestaDTO> obtenerEncuestaPorTokenEncuesta(HttpServletRequest httpServletRequest, BodyEncriptado bodyEncriptado);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data EncuestaDTO
     *
     * @param httpServletRequest request peticion.
     * @return RespuestaPorDefectoAuditoria<EncuestaDTO>
     */
    RespuestaPorDefectoAuditoria<EncuestaDTO> obtenerEvaluacionPorTokenEncabezado(HttpServletRequest httpServletRequest, BodyEncriptado bodyEncriptado);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data EncuestaDTO
     *
     * @param httpServletRequest request peticion.
     * @return RespuestaPorDefectoAuditoria<EncuestaDTO>
     */
    RespuestaPorDefectoAuditoria<EncuestaDTO> obtenerEncuestaPorId(HttpServletRequest httpServletRequest,
            EncuestaDTO encuestaDTO);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data EncuestaDTO si la
     * encuesta se creo con exito
     *
     * @param httpServletRequest request peticion.
     * @param encuestaDTO        objeto encuesta dto.
     * @return RespuestaPorDefectoAuditoria<EncuestaDTO>
     */
    RespuestaPorDefectoAuditoria<EncuestaDTO> crearEncuesta(HttpServletRequest httpServletRequest,
            EncuestaDTO encuestaDTO);
    
    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data EncuestaDTO si la
     * encuesta se creo con exito
     *
     * @param httpServletRequest request peticion.
     * @param bodyEncriptado        objeto BodyEncriptado.
     * @return RespuestaPorDefectoAuditoria<EncuestaDTO>
     */
    RespuestaPorDefectoAuditoria<Boolean> crearEvaluacion(HttpServletRequest httpServletRequest,
                                                          BodyEncriptado bodyEncriptado);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data EncuestaDTO si la
     * empresa se creo con exito
     *
     * @param httpServletRequest request peticion.
     * @param bodyEncriptado        objeto encuesta dto.
     * @return RespuestaPorDefectoAuditoria<EncuestaDTO>
     */
    RespuestaPorDefectoAuditoria<Boolean> actualizarEncuesta(HttpServletRequest httpServletRequest,
                                                             BodyEncriptado bodyEncriptado);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data Boolean si la
     * encuesta se remueve con exito
     *
     * @param httpServletRequest request peticion.
     * @param bodyEncriptado        objeto encuesta dto.
     * @return RespuestaPorDefectoAuditoria<EncuestaDTO>
     */
    RespuestaPorDefectoAuditoria<Boolean> removerEncuesta(HttpServletRequest httpServletRequest,
                                                          BodyEncriptado bodyEncriptado);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data Boolean si la
     * evaluacion se remueve con exito
     *
     * @param httpServletRequest request peticion.
     * @param bodyEncriptado        objeto encabezado dto.
     * @return RespuestaPorDefectoAuditoria<Boolean>
     */
    RespuestaPorDefectoAuditoria<Boolean> removerEvaluacion(HttpServletRequest httpServletRequest,
                                                          BodyEncriptado bodyEncriptado);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data Boolean si el documento se sube con exito
     *
     * @param httpServletRequest request peticion.
     * @param bodyEncriptado         objeto BodyEncriptado.
     * @return RespuestaPorDefectoAuditoria<Boolean>
     */
    RespuestaPorDefectoAuditoria<Boolean> subirDocumento(HttpServletRequest httpServletRequest,
                                                              MultipartFile[] multipartFile,
                                                              BodyEncriptado bodyEncriptado);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data PaginacionResponse<DocumentoDTO>
     *
     * @param httpServletRequest request peticion.
     * @param bodyEncriptado         objeto BodyEncriptado.
     * @return RespuestaPorDefectoAuditoria<Boolean>
     */
    RespuestaPorDefectoAuditoria<PaginacionResponse<DocumentoDTO>> obtenerDocumentos(HttpServletRequest httpServletRequest,
                                                                                     BodyEncriptado bodyEncriptado);

    // endregion
}

File: pronacej-front/src/app/modules/general/services/encuesta.service.ts
import { Injectable } from '@angular/core';
import { BodyEncriptado } from 'app/core/model/both/bodyEncriptado.model';
import { DocumentoDTO } from 'app/core/model/both/DocumentoDTO.model';
import { EncabezadoDTO } from 'app/core/model/both/encuesta/encabezadoDTO.model';
import { EncuestaDTO } from 'app/core/model/both/encuesta/encuestaDTO.model';
import { EvaluacionDocumentoRequest } from 'app/core/model/request/general/EvaluacionDocumentoRequest.model';
import { PaginacionRequest } from 'app/core/model/request/PaginacionRequest.model';
import { PaginacionResponse } from 'app/core/model/response/PaginacionResponse.model';
import { RespuestaPorDefecto } from 'app/core/model/response/RespuestaPorDefecto.model';
import { BackendService } from 'app/core/services/backend.service';
import { Observable, Subscriber } from 'rxjs';

@Injectable({
    providedIn: 'root',
})
export class EncuestaService {
    private path = '/encuesta';

    constructor(private backendService: BackendService) {}

    obtenerListaEncuestas(
        paginacionRequest: PaginacionRequest,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<PaginacionResponse<EncuestaDTO>>> {
        let endPoint = this.path + '/obtenerListaEncuestas';

        return this.backendService.postFinal(
            endPoint,
            paginacionRequest,
            nemonicoMenu
        );
    }

    obtenerEncuestas(
        encuestaDTO: EncuestaDTO,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<EncuestaDTO[]>> {
        let endPoint = this.path + '/obtenerEncuestas';
        return this.backendService.postFinal(
            endPoint,
            encuestaDTO,
            nemonicoMenu
        );
    }

    obtenerEvaluacionesPorFichaIdentificacion(
        paginacionRequest: PaginacionRequest,
        nemonicoCategoria: string,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<PaginacionResponse<EncabezadoDTO>>> {
        let endPoint =
            this.path +
            '/obtenerEvaluacionesPorFichaIdentificacion?nemonicoCategoria=' +
            nemonicoCategoria;
        return this.backendService.postFinal(
            endPoint,
            paginacionRequest,
            nemonicoMenu
        );
    }

    obtenerEvaluacionesPorNemonicoEncuesta(
        paginacionRequest: PaginacionRequest,
        nemonicoEncuesta: string,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<PaginacionResponse<EncabezadoDTO>>> {
        let endPoint =
            this.path +
            '/obtenerEvaluacionesPorNemonicoEncuesta?nemonicoEncuesta=' +
            nemonicoEncuesta;
        return this.backendService.postFinal(
            endPoint,
            paginacionRequest,
            nemonicoMenu
        );
    }

    obtenerEvaluacionesPorNemonicoCategoria(
        paginacionRequest: PaginacionRequest,
        nemonicoCentro: string,
        nemonicosCategoria: string[],
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<PaginacionResponse<EncabezadoDTO>>> {
        const listaSerializada = nemonicosCategoria.join(',');
        const endPoint =
            `${this.path}/obtenerEvaluacionesPorNemonicoCategoria` +
            `?nemonicoCentro=${nemonicoCentro}` +
            `&nemonicosCategoria=${encodeURIComponent(listaSerializada)}`;
        return this.backendService.postFinal(
            endPoint,
            paginacionRequest,
            nemonicoMenu
        );
    }

    obtenerEncuestaPorTokenEncuesta(
        encabezadoDTO: EncabezadoDTO,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<EncuestaDTO>> {
        let endPoint = this.path + '/obtenerEncuestaPorTokenEncuesta';
        return this.backendService.postFinal(
            endPoint,
            encabezadoDTO,
            nemonicoMenu
        );
    }

    obtenerEvaluacionPorTokenEncabezado(
        encabezadoDTO: EncabezadoDTO,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<EncuestaDTO>> {
        let endPoint = this.path + '/obtenerEvaluacionPorTokenEncabezado';
        return this.backendService.postFinal(
            endPoint,
            encabezadoDTO,
            nemonicoMenu
        );
    }

    crearEvaluacion(
        encabezadoDTO: EncabezadoDTO,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<Boolean>> {
        let endPoint = this.path + '/crearEvaluacion';
        return this.backendService.postFinal(
            endPoint,
            encabezadoDTO,
            nemonicoMenu
        );
    }

    crearEncuesta(
        encuestaDTO: EncuestaDTO,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<EncuestaDTO>> {
        let endPoint = this.path + '/crearEncuesta';
        return this.backendService.postFinal(
            endPoint,
            encuestaDTO,
            nemonicoMenu
        );
    }

    actualizarEncuesta(
        encuestaDTO: EncuestaDTO,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<EncuestaDTO>> {
        let endPoint = this.path + '/actualizarEncuesta';
        return this.backendService.postFinal(
            endPoint,
            encuestaDTO,
            nemonicoMenu
        );
    }

    eliminarEncuesta(
        encuestaDTO: EncuestaDTO,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<boolean>> {
        let endPoint = this.path + '/removerEncuesta';
        return this.backendService.postFinal(
            endPoint,
            encuestaDTO,
            nemonicoMenu
        );
    }

    eliminarEvaluacion(
        encabezadoDTO: EncabezadoDTO,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<boolean>> {
        let endPoint = this.path + '/removerEvaluacion';
        return this.backendService.postFinal(
            endPoint,
            encabezadoDTO,
            nemonicoMenu
        );
    }

    /**
     * Sube un informe firmado
     *
     * @param encabezadoDTO InformeDTO
     * @param archivo File
     * @param nemonicoMenu string nemonico menu
     *
     * @return Observable<RespuestaPorDefecto<Boolean>>
     */
    subirDocumentos(
        encabezadoDTO: EncabezadoDTO,
        files: File[],
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<Boolean>> {
        let endPoint = this.path + '/subirDocumento';

        let formData = new FormData();
        if (files != null) {
            for (let file of files) {
                formData.append('documentos', file);
            }
        }

        return new Observable(
            (subs: Subscriber<RespuestaPorDefecto<Boolean>>) => {
                this.backendService
                    .crearBodyEncriptado(encabezadoDTO)
                    .then((bodyEncriptado) => {
                        formData.append('body', JSON.stringify(bodyEncriptado));
                        this.backendService
                            .postFormDataBodyEncriptado2(
                                endPoint,
                                formData,
                                nemonicoMenu
                            )
                            .subscribe({
                                next: async (body: BodyEncriptado) => {
                                    let resp =
                                        await this.backendService.desencriptarBdyEncriptado<
                                            RespuestaPorDefecto<Boolean>
                                        >(body);
                                    subs.next(resp);
                                    subs.complete();
                                },
                                error: (error: any) => {
                                    subs.error(error);
                                    subs.complete();
                                },
                            });
                    })
                    .catch((error: any) => {
                        subs.error(error);
                        subs.complete();
                    });
            }
        );
    }

    obtenerDocumentos(
        request: EvaluacionDocumentoRequest,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<PaginacionResponse<DocumentoDTO>>> {
        let endPoint = this.path + '/obtenerDocumentos';
        return this.backendService.postFinal(endPoint, request, nemonicoMenu);
    }
}
